### ISSUE TITLE
Uncalled function checks only works with single conditional

### ISSUE THREAD

**Issue by mjbvz** [microsoft/TypeScript#35584]
**TypeScript Version:**  3.8.0-dev.20191207

**Search Terms:**

**Code**
With strict null checks enabled:

```ts
import * as fs from 'fs';

fs.stat('/path/to/file', function(err, stats) {
    if (stats.isFile || stats.isDirectory) {

    }
});
```

**Expected behavior:**
Accesses to `isFile` and `isDirectory` are reported as errors since these are actually functions

**Actual behavior:**
No errors. 

Errors are only reported when there is a single conditional

**Playground Link:** 

**Related Issues:**

**Comment by RyanCavanaugh** [comment]
This is the intended scoping of the check to reduce over-reported errors.

**Comment by clarkio** [comment]
@RyanCavanaugh I can definitely appreciate the reduction in over-reported errors but do you think not having any error at all seems to defeat the purpose? Could there be a way to mention it once for a particular block of code?

**Comment by RyanCavanaugh** [comment]
We don't have "Just so you know" warnings and don't intend to add them

**Comment by clarkio** [comment]
I think I may be miscommunicating the scenario. I'm hoping to receive the error for the one instance even if it is included with other conditionals.

For example here we accidentally don't call `isDirectory` as a function while having some other conditional with it.
```typescript
import * as fs from 'fs';

fs.stat('/path/to/file', function(err, stats) {
    if (stats.isDirectory && !err) {

    }
});
```

At the moment if we were to do this the error around `stats.isDirectory` is not reported at all but if we remove the `!err` conditional it will report as expected.

Do you think it should still be reported in this scenario?

**Comment by DanielRosenwasser** [comment]
> This is the intended scoping of the check to reduce over-reported errors.

I may have missed any checks for whether this comes up in conditionals, `&&`, and `||` - did we see too many problems on those? My hunch is that that'd be nothing but goodness.

**Comment by RyanCavanaugh** [comment]
We can try expanding the set of things checked and see what comes up.

**Comment by dragomirtitian** [comment]
@DanielRosenwasser How deep do we want to go here? Any sort of arbitrary expression with `boolean` operators should be ok ? I think the interesting bit is when the condition both checks and uses the function:

```ts

interface StatsBase<T> {
    isDirectory(): boolean;
    time: number
}

function A(err: any, stats: StatsBase<any>) {
    if (stats.isDirectory || stats.isDirectory()) { // should be error ?
        
    }
}

function B(err: any, stats: StatsBase<any>) {
    if (stats.isDirectory && stats.isDirectory()) { // should be ok 
    }
}


function C(err: any, stats: StatsBase<any>) {
    if (!stats.isDirectory || stats.isDirectory()) { // should be ok 
    }
}

``` 
 [Playground Link](https://www.typescriptlang.org/play/?ssl=23&ssc=1&pln=1&pc=1#code/LAKAlgdgLgpgTgMwIYGMYAIDKUlQM4BCSeMAPACoB86A3qOg+mHgCJhwwpQD2cAngAoAlAC50AI27cANjCQQA3PUZQwAWxhiIAVzXj4oAL6hQCbRC5huEdAEEB8OGPl8ANOjw58Y7LkLEyF0ohWmUGMAR0AU8-ADpmNg4uXj50AB80jy88eNZ2Th5+YRCadAB6Mo8AC25taQATCQxHXnQAfjDGBk7jEF7Tc0trdAIHOCd0F3cY7yxsohJSIJLOiKiZnIT85P50ADI9rLitpMLBIRLyyrwausb9dG4Aa3QeoxMQAYtVYYBhMYmUyOs18+AWgQgfGCoRAXTWAgAhBtcokCil0plkSc0UULrQrtVag0mo8Xm8+kA)

**Comment by nityatammana** [comment]
Can @mshivaku99 and I work on this (we have a class assignment to help fix a bug in an open-source project)

**Comment by mykolao-wix** [comment]
@sandersn
@DanielRosenwasser is it decided if this issue needs to be resolved?

I am doing open-source for this week and would like to help resolve `good-first-issue`/`GraceHopperOSD`'s issues

**Comment by sandersn** [comment]
@mykolao-wix yep anything in a milestone (such as Backlog) has been accepted. In this case, Experimentation Needed+Suggestion means that the proposed solution might not be workable, but we want to find out by looking at a real implementation.

**Comment by mkloouo** [comment]
@sandersn is there anyone I can contact to ask `src/compiler/checker.ts` API questions?

**Comment by evmar** [comment]
We had a variant of this come up at Google in code like

```
if (!foo.hasBar) {  // oops, forgot to call the function
  ...
} else {
  // assume foo.hasBar is true
}
```

I think the broader fix might be something around "expression is always true" when you do a boolean check on a non-optional method.

[Playground repro](https://www.typescriptlang.org/play?ts=4.1.3&ssl=9&ssc=2&pln=2&pc=1#code/FAEwpgxgNghgTmABAMwK4DsIBcCWB7dRHAZwDE88AKASgC5EAjCqMGdAbmGB2UUpPJ5qiAN7BEExAHopiMAA8ADpCxgQcuHDxwANIgDmeLBq1xgAX269KAQgEVhYydNkLl2NSe17Dx9Hi8zcyA)
```
declare function isFoo(): boolean;

if (isFoo) {
    // expected error, got error
}
if (!isFoo) {
    // expected error, got no error
}
```

**Comment by jonhue** [comment]
Just opened a pull request (https://github.com/microsoft/TypeScript/pull/42835) that attempts to address the two problems mentioned in this thread. Feel free to check it out 🙂
